modules/logger/logger.go

```diff
@@ -7,36 +7,135 @@ import (
 	"os"
 	"runtime"
 	"strings"
+	gotesting "testing"
 	"time"
 
 	"github.com/gruntwork-io/terratest/modules/testing"
 )
 
+var (
+	// DefaultLogf will be used if the caller uses the function Logf, where on the first call to that function, a sane
+	// default logging function will be set. We keep that in a global variable so that we don't need to define it on
+	// every call to Logf.
+	// DefaultLogf is the default logger that is used if no other logger should be provided. It uses the TerratestLogger
+	// to log messages. This can be overwritten to change the logging globally.
+	Default = New(terratestLogger{})
+	// Discard discards all logging.
+	Discard = New(discardLogger{})
+	// Terratest logs the given format and arguments, formatted using fmt.Sprintf, to stdout, along with a timestamp and
+	// information about what test and file is doing the logging. Before Go 1.14, this is an alternative to t.Logf as it
+	// logs to stdout immediately, rather than buffering all log output and only displaying it at the very end of the test.
+	// This is useful because:
+	//
+	// 1. It allows you to iterate faster locally, as you get feedback on whether your code changes are working as expected
+	//    right away, rather than at the very end of the test run.
+	//
+	// 2. If you have a bug in your code that causes a test to never complete or if the test code crashes, t.Logf would
+	//    show you no log output whatsoever, making debugging very hard, where as this method will show you all the log
+	//    output available.
+	//
+	// 3. If you have a test that takes a long time to complete, some CI systems will kill the test suite prematurely
+	//    because there is no log output with t.Logf (e.g., CircleCI kills tests after 10 minutes of no log output). With
+	//    this log method, you get log output continuously.
+	//
+	Terratest = New(terratestLogger{})
+	// TestingT can be used to use Go's testing.T to log. If this is used, but no testing.T is provided, it will fallback
+	// to Default.
+	TestingT = New(testingT{})
+)
+
+type TestLogger interface {
+	Logf(t testing.TestingT, format string, args ...interface{})
+}
+
+type Logger struct {
+	l TestLogger
+}
+
+func New(l TestLogger) *Logger {
+	return &Logger{
+		l,
+	}
+}
+
+func (l *Logger) Logf(t testing.TestingT, format string, args ...interface{}) {
+	if tt, ok := t.(helper); ok {
+		tt.Helper()
+	}
+
+	if l == nil || l.l == nil {
```

Comment:
I don't think `l` will ever be `nil`, given that you can't do `nil.Logf`.
```suggestion
	if l.l == nil {
```

Reply:
Typing this on my phone, will have a look at everything tomorrow.

However, you can do this; which is also why I don‘t check for nil anywhere.
Basically, what we‘re doing is:

<pre>
var l *Logger // is now a nil pointer of type *Logger
l.Logf // actually translates to Logf(nil, ...)
</pre>

That‘s why I‘m checking for nil here, as an nil pointer can then be made useful as the default.

Reply:
https://play.golang.org/p/isXuh90a3E1

Reply:
Ahhh that is neat. I didn't know that. Can you add a comment to reference this, since it's not something that is common that we do?

Reply:
Yup, creates a clean and non-dangerous API 😄 
Added a comment.